### PR TITLE
[abstractComponentContainer] Can now merge groups that are already anchored

### LINKED ISSUE #1823: Can't merge Component that's already anchored
Repro: http://jsfiddle.net/0r93afbf/
Trying to add the label again produces the following message:
```
Uncaught Error: Can't presently merge a component that's already been anchored
```

Expected: adding a `Component` to a new `Group` should work even if the Component is already rendered.

### PR DESCRIPTION
**Pull request by aicioara**
Changelog:
- No longer setting the parent field, but rather have a method for doing that, which also detaches the component from its old parent (if there's the case)
- No longer throwing an error if the component was already setup or anchored, as we can now do it
- Throwing an error if we are trying to add `null` to a table cell

Closes #1823 

Before: http://jsfiddle.net/0r93afbf/
After: http://jsfiddle.net/0r93afbf/5/

Fiddle explanation: Click on the button. If it works correctly, the label "AWESOME" should be taken from the x-Axis and put inside the graph. The default position is center (vertically and horizontally)

### CODE REVIEW

**Comment by aicioara** [comment]
.

File: src/components/abstractComponent.ts [hunk 1]
```diff
@@ -491,6 +488,11 @@ export module Component {
       return this;
     }
 
+    public _setParent(parent: AbstractComponentContainer) {
```
Comment (jtlan) [inline src/components/abstractComponent.ts]: Since we have this method now, please make `_parent` private or protected.

File: test/core/componentGroupTests.ts [hunk 2]
```diff
@@ -185,6 +185,48 @@ describe("ComponentGroups", () => {
...
+
+    it("can move components to other groups after anchoring", () => {
+      var svg = generateSVG();
+
+      var xScale = new Plottable.Scale.Category();
+      xScale.domain(["A", "B", "C"]);
+      var xAxis = new Plottable.Axis.Category(xScale, "bottom");
+      var yScale = new Plottable.Scale.Linear();
+      var yAxis = new Plottable.Axis.Numeric(yScale, "left");
+      var chart = new Plottable.Component.Table([
```
Comment (jtlan) [inline test/core/componentGroupTests.ts]: You shouldn't have to use a `Table` in this test; simply create two `Group`s in different `<svg>`s and move a `Component` between them after rendering.

File: src/components/table.ts [hunk 3]
```diff
@@ -51,14 +51,16 @@ export module Component {
       this.classed("table", true);
       rows.forEach((row, rowIndex) => {
         row.forEach((component, colIndex) => {
-          this.addComponent(rowIndex, colIndex, component);
+          if (component != null) {
```
Comment (jtlan) [inline src/components/table.ts]: Is there a test associated with this change?

**Comment by jtlan** [comment]
In general, when writing tests, try to use as few parts of the code not under test as possible. In this case, for example, your test includes a `Table` and two `Axes` -- but the bug in question can manifest using only a pair of `Group`s (the `Component` being tested) and an `AbstractComponent` (the parent class).

You can think of a test as though it were the proof of a theorem -- you want to base them either on axioms (in this analogy, code assumed to work, like fundamental parts of Javascript) or other theorems you've already proven, preferably those the shortest distance from an axiom (in this analogy, other tested parts of the code, higher up the inheritance chain if possible).

File: src/components/table.ts (hunk 3 above)
Comment (aicioara) [inline src/components/table.ts]: This is just an optimization. The case is caught later in the abstractComponentContainer._addComponent(). Not sure what you mean with testing

**Comment by aicioara** [comment]
Fixed the test.

File: src/components/table.ts (hunk 3 above)
Comment (jtlan) [inline src/components/table.ts]: Maybe we should test the behavior of `this.addComponent(row, col, null)` if the cell is already occupied.

File: test/core/componentGroupTests.ts [hunk 4]
```diff
@@ -185,6 +185,46 @@ describe("ComponentGroups", () => {
...
+      );
+
+      assert.strictEqual(cg2.components().length, 1,
+        "Second group should have 1 component after movement");
+
+      assert.strictEqual(cg1.components().length, 0,
+        "First group should have no components after movement");
+
+      assert.strictEqual(c._parent(), cg2,
+        "component's parent after movement should be the group 1"
```
Comment (jtlan) [inline test/core/componentGroupTests.ts]: Should be group 2?

**Comment by jtlan** [comment]
Minor test error, and we probably should have a test for `Table`'s `addComponent()` method.

File: src/components/table.ts (hunk 3 above)
Comment (aicioara) [inline src/components/table.ts]: Done

**Comment by aicioara** [comment]
All suggested changes added. Waiting for compilation.

File: src/components/table.ts [hunk 5]
```diff
@@ -83,8 +85,7 @@ export module Component {
       var currentComponent = this._rows[row] && this._rows[row][col];
 
       if (currentComponent) {
-        currentComponent.detach();
-        component = component.above(currentComponent);
+        component = currentComponent.below(component);
```
Comment (jtlan) [inline src/components/table.ts]: Why this change?

**Comment by aicioara** [comment]
Discussion with @jtlan and @bluong concluded that there are 3 options when adding null to a table cell where there is something
- Throw an Error - We cannot add null to a table cell. Pro for this was that the user is notified straight away that something wrong has happened. Code does not fail later on unexpectedly
- Do not do anything - fail silently. Pro for this would be that user might have a list of components, some of which are null and wants to add them to the cell
- Create a group containing only 1 element (the element which was originally in the cell). Pro for this was that user expects a group, even though he added a null

Went for option 1

Solved that. Adding `null` to a table cell now throws an exception. Tests changed accordingly.

File: test/core/tableTests.ts [hunk 6]
```diff
@@ -106,6 +106,13 @@ describe("Tables", () => {
     assert.equal(components[2], c3, "The Component was added to the existing Group");
   });
 
+  it("add null to a table cell where there was a group should throw an error", () => {
```
Comment (jtlan) [inline test/core/tableTests.ts]: "where there was a group" is no longer necessary in this description.

**Comment by jtlan** [comment]
One minor thing.

**Comment by crmorford** [comment]
should i be able to steal back an anchored plot and render it to its own svg? http://jsfiddle.net/ck1psjL3/22/, lines 26&28

**Comment by crmorford** [comment]
Update on last comment. Some sort of bug. 
http://jsfiddle.net/ck1psjL3/25/ (i can't un-anchor a plot and render it to a different size svg correctly)
vs http://jsfiddle.net/ck1psjL3/24/ (i can render a plot back and forth between svgs no problem)

**Comment by aicioara** [comment]
Working on that

**Comment by aicioara** [comment]
Fixed:
Note the detach http://jsfiddle.net/ck1psjL3/27/
@jtlan do you want this in code as well? See my commit.

**Comment by jtlan** [comment]
Yes, a `Component` that gets sent anywhere should disconnect itself from its parent.

**Comment by aicioara** [comment]
Fixed

File: test/core/componentTests.ts [hunk 7]
```diff
@@ -403,6 +403,33 @@ describe("Component behavior", () => {
     svg.remove();
   });
 
+  it("rendering to a new svg detaches the component", () => {
```
Comment (jtlan) [inline test/core/componentTests.ts]: I think properly this test belongs somewhere else, but I think we'll leave it for now.

Comment (aicioara) [inline test/core/componentTests.ts]: I actually gave it a thought. I put it in the same category with the other detach() methods on the component.
